libevmasm/Assembly.cpp

```diff
@@ -297,6 +503,142 @@ Json::Value Assembly::assemblyJSON(std::map<std::string, unsigned> const& _sourc
 	return root;
 }
 
+std::pair<std::shared_ptr<Assembly>, std::vector<std::string>> Assembly::fromJSON(
+	Json::Value const& _json,
+	std::vector<std::string> const& _sourceList,
+	size_t _level
+)
+{
+	solRequire(_json.isObject(), AssemblyImportException, "Supplied JSON is not an object.");
+	static std::set<std::string> const validMembers{".code", ".data", ".auxdata", "sourceList"};
+	for (std::string const& attribute: _json.getMemberNames())
+		solRequire(validMembers.count(attribute), AssemblyImportException, "Unknown attribute '" + attribute + "'.");
+
+	if (_level == 0)
+	{
+		if (_json.isMember("sourceList"))
+		{
+			solRequire(_json["sourceList"].isArray(), AssemblyImportException, "Optional member 'sourceList' is not an array.");
+			for (Json::Value const& sourceName: _json["sourceList"])
+				solRequire(sourceName.isString(), AssemblyImportException, "The 'sourceList' array contains an item that is not a string.");
+		}
+	}
```

Comment:
```suggestion
if (_json.isMember("sourceList"))
{
	solAssert(_level == 0);
	solRequire(_json["sourceList"].isArray(), AssemblyImportException, "Optional member 'sourceList' is not an array.");
	for (Json::Value const& sourceName: _json["sourceList"])
		solRequire(sourceName.isString(), AssemblyImportException, "The 'sourceList' array contains an item that is not a string.");
}
```

My understanding is that if `sourceList` is a member, then `_level` must be zero (see lines 535-537), so this `_level == 0` check seems unnecessary.

Reply:
Your version will fail on an assert if a `sourceList` is found at a level other than `0`.

It was written like it was because there are 3 cases we need to consider and validate:
- `"sourceList"` present and `_level == 0`: `"sourceList"` must have the right structure.
- `"sourceList"` present and `_level > 0`: nothing to validate.
- `_level > 0` : `"sourceList"` must not be present.

Reply:
Ah, OK, so there is in fact a third option :+1: